File: src/main/scala/com/lightbend/akka/sample/SampleHelpers.scala
/*
 * Copyright (C) 2017 Lightbend Inc. <http://www.typesafe.com>
 */
package com.lightbend.akka.sample

import akka.actor.ActorSystem
import akka.cluster.{Cluster, MemberStatus}

import scala.concurrent.duration.FiniteDuration

object SampleHelpers {

  /**
   * Block the calling thread until all the given nodes have reached Up and that is seen from all nodes.
   * In a real system (tm) you should have no use for such a thing but instead use an actor listening for
   * cluster membership events. This is here to keep samples brief and easy to follow.
   */
  def waitForAllNodesUp(nodes: ActorSystem*): Unit = {
    if (nodes.exists(node => Cluster(node).state.members.count(_.status == MemberStatus.Up) != nodes.size)) {
      Thread.sleep(250)
      waitForAllNodesUp(nodes:_*)
    } else ()
  }

  /**
   * Block the calling thread for the given time period. In real system (tm) built with
   * Akka you should never have anything like this - it is just for making the samples
   * easier to follow and understand.
   */
  def wait(d: FiniteDuration): Unit = {
    Thread.sleep(d.toMillis)
  }

}

File: src/main/java/com/ligthbend/akka/cluster/java/sample/SampleHelpers.java
/**
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */
package com.ligthbend.akka.cluster.java.sample;

import akka.actor.ActorSystem;
import akka.cluster.Cluster;
import akka.cluster.Member;
import akka.cluster.MemberStatus;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class SampleHelpers {

  public static Config portConfig(int port) {
    return ConfigFactory.parseString("akka.remote.artery.canonical.port = " + port);
  }




  /**
   * Block the calling thread until all the given nodes have reached Up and that is seen from all nodes.
   * In a real system (tm) you should have no use for such a thing but instead use an actor listening for
   * cluster membership events. This is here to keep samples brief and easy to follow.
   */
  public static void waitForAllNodesUp(ActorSystem... nodes) throws InterruptedException {
    boolean allNodesUp = false;
    final List<ActorSystem> nodeList = Arrays.asList(nodes);
    do {
      allNodesUp = nodeList.stream().anyMatch(system -> {
        int upNodes = 0;
        Iterable<Member> members = Cluster.get(system).state().getMembers();
        for (Member member : members) {
          if (member.status().equals(MemberStatus.up()))
            upNodes++;
        }
        return upNodes != nodeList.size();
      });
      // don't ever block like this in actual code
      Thread.sleep(250);
    } while (!allNodesUp);
  }

  /**
   * Block the calling thread for the given time period. In real system (tm) built with
   * Akka you should NEVER have anything like this - it is just for making the samples
   * easier to follow and understand.
   */
  public static void wait(Duration d) {
    try {
      Thread.sleep(d.toMillis());
    } catch (InterruptedException ex) {
      throw new RuntimeException(ex);
    }
  }
}
